packages/typescript-client/src/client.ts

```
@@ -319,10 +318,18 @@ export class ShapeStream<T extends Row = Row>
     return !this.isUpToDate
   }
 
-  #publish(messages: Message<T>[]) {
-    this.#subscribers.forEach(([subscriber, _]) => {
-      subscriber.process(messages)
-    })
+  async #publish(messages: Message<T>[]): Promise<void> {
+    await Promise.all(
+      Array.from(this.#subscribers.values()).map(async ([callback, __]) => {
+        try {
+          await callback(messages)
+        } catch (err) {
+          queueMicrotask(() => {
+            throw err
```

how would a user catch this error?

---

With a global window.onerror or unhandledpromisererejection.
Key thing is that it will show in console and get logged by tools such as Sentry.

---

Yeah I guess that works. This is just an escape hatch. Generally you'd catch and handle the errors in the actual subscriber code.

---

Yep exactly. It's to stop them from being silently eaten.